javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -672,13 +672,27 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
 
   def multiply(multiplicand: BigDecimal): BigDecimal = {
     val newScale = this._scale.toLong + multiplicand._scale
+
+    @inline
+    def bigMultiply: BigDecimal = {
+      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
+      new BigDecimal(unscaled, safeLongToInt(newScale))
+    }
+
+    import java.lang.{Long => JLong}
     if (this.isZero || multiplicand.isZero) {
       zeroScaledBy(newScale)
-    } else if (this._bitLength + multiplicand._bitLength < 64) {
-      valueOf(this._smallValue * multiplicand._smallValue, safeLongToInt(newScale))
+    } else if (this._bitLength + multiplicand._bitLength >= 64) {
+      bigMultiply
     } else {
-      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
-      new BigDecimal(unscaled, safeLongToInt(newScale))
+      val x = this._smallValue
+      val y = multiplicand._smallValue
+      val z = x * y
+      if (JLong.signum(z) == JLong.signum(x) * JLong.signum(y)) {
```

Comment:
So my original description was wrong -- it really has to be something like `MinValue * MinValue * 2` to trigger the bug, which means that there are at least two cases to check instead of one.

If you would prefer to do it that way I can go ahead and save a `Long` constant and then do two tests to catch it. I think doing the `Long` multiplication rather than those two tests ends up being a bit more efficient, but either way works for me.